src/cswinrt/strings/additions/Windows.Storage.Streams/WindowsRuntimeMarshal.cs

```diff
@@ -0,0 +1,57 @@
+namespace System.Runtime.InteropServices.WindowsRuntime
+{
+    using System.Diagnostics;
+    using System.Diagnostics.CodeAnalysis;
+    using System.IO;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using global::Windows.Foundation;
+    using global::Windows.Storage.Streams;
+    using WinRT;
+    /// <summary>
+    /// Contains extension methods that expose operations on WinRT <code>Windows.Foundation.IBuffer</code>.
+    /// </summary>
+#if EMBED
+    internal
+#else
+    public
+#endif
+    static partial class WindowsRuntimeMarshal
+    {
+        public static unsafe bool TryGetDataUnsafe(IBuffer buffer, out IntPtr dataPtr)
+        {
+            if (buffer == null)
+            {
+                dataPtr = IntPtr.Zero;
+                return false;
+            }
+
+            if (ComWrappersSupport.TryUnwrapObject(buffer, out var unwrapped) &&
+                unwrapped.TryAs(global::ABI.Windows.Storage.Streams.IBufferByteAccessMethods.IID, out IntPtr ThisPtr) >= 0)
+            {
+                try
+                {
+                    IntPtr __retval = default;
+                    global::WinRT.ExceptionHelpers.ThrowExceptionForHR((*(delegate* unmanaged[Stdcall]<IntPtr, IntPtr*, int>**)ThisPtr)[3](ThisPtr, &__retval));

```

Comment:
Why the roundtrip to a managed exception? We could just check the HRESULT instead.

Reply:
We aren't roundtripping. We still want to throw if the call fails.

Reply:
Oops, sorry I misread, nevermind 😅